### ISSUE TITLE
[🐛 Bug]: `@vercel/og` doesn't work

### ISSUE THREAD

**Issue by w00fz** [cloudflare/next-on-pages#39]
I'm having issues getting my project to successfully deploy on Cloudflare because I am relying on a dependency (`@vercel/og`) which uses `satori/wasm`.

The Vercel building process goes fine but when `next-on-pages` gets to build the final worker bundle , `esbuild` fails with error:

```bash
✘ [ERROR] Could not resolve "/wasm/wasm_53adb396180a95eb122a0218c93e3d6474b1a27e.wasm"
```

Any idea on how this could get resolved?

**Comment by GregBrimble** [comment]
We don't currently support Wasm in Pages Functions, but we hope to add this soon. We'll update this issue when we get around to this.

**Comment by PerchunPak** [comment]
Is there any workaround/update?

**Comment by dario-piotrowicz** [comment]
### Note
This issue is blocked by: https://github.com/cloudflare/next-on-pages/issues/202 and https://github.com/cloudflare/pages-plugins/issues/30

Most likely solving the two issues above should in turn also solve this one

**Comment by leonlarsson** [comment]
Ahh okay. Thanks :)

**Comment by dario-piotrowicz** [comment]
No problem, sorry this is taking so long :sweat:, I'll definitely look into #202 soon (likely this or next week), I'll try to also get the pages-plugins one solved soon-ish as well :slightly_smiling_face:

**Comment by dario-piotrowicz** [comment]
Sorry I misunderstood things, we don't necessarily want to force developers to use the vercel-og pages plugin, so that is not blocking this

`@vercel/og` anyways isn't yet currently working on next-on-pages, so that is indeed something we need to fix

in any case I'm going to update the issue title as it is now unrelated to `wasm` (and also that it contains `@vercel/og` so that people having troubles with it can more easily notice this issue)

**Comment by lucaswalter** [comment]
Have been following a couple of different issue thread but the current stat is unclear to me. At the moment, is there any workaround or way to use either `@vercel/og` or `@cloudflare/pages-plugin-vercel-og/api`?

**Comment by dario-piotrowicz** [comment]
Hi @lucaswalter, no I'm sorry, I had a quick look some time ago but haven't dug much into it since, so it's still unresolved 😓

**Comment by lucaswalter** [comment]
Thanks for the update @dario-piotrowicz — Do you know what the main blocker is for this? I would like to make use of this for my projection on Pages, but am not sure where else to follow for status updates.

**Comment by dario-piotrowicz** [comment]
@lucaswalter as far I know there actually isn't a real blocker, it's just a matter of finding the time to dig into what the issue  with `vercel/og` actually is.

it's been a few weeks so I don't remember precisely, but I think there was something Vercel does that we need to reverse engineer and fix/adapt to our runtime.

Sorry, I haven't put too much priority into this since being related to a Vercel library so it's not a core issue with Next per se, I'll try to dig into this soon (this or next week) 🙇

PS: I'll post status updates here

**Comment by lucaswalter** [comment]
Understood, thanks a ton @dario-piotrowicz

**Comment by dario-piotrowicz** [comment]
No problem @lucaswalter 🙂 
 
 Check out #325 we're very close! 😄 (feel free to try out the [next-on-pages prerelease](https://github.com/cloudflare/next-on-pages/pull/325#issuecomment-1604457773) and let me know if it works for you if you want)

**Comment by lucaswalter** [comment]
Thanks @dario-piotrowicz — Will keep following along

**Comment by dario-piotrowicz** [comment]
@lucaswalter I've merged the PR so `@vercel/og` should now work 🙂 

See the example I added in the PR's description: https://a9aeda25.next-on-pages-test-5h3.pages.dev/

As you can see ([third link](https://a9aeda25.next-on-pages-test-5h3.pages.dev/api/og-custom)) now `@vercel/og` should work, unfortunately it seems like in certain cases
we can get an error ([second link](https://a9aeda25.next-on-pages-test-5h3.pages.dev/api/og)), I believe that that is a runtime bug and I will check with the runtime team about it,
so let's keep this issue open a bit longer, so I will let you know here when that is also sorted out

**Comment by w00fz** [comment]
This is awesome, thank you!

**Comment by dario-piotrowicz** [comment]
I'm closing the issue since I could not reproduce it using the [demo app](https://github.com/dario-piotrowicz/next-apps-for-testing/tree/master/apps/simple-wasm-pages-13.3.1) anymore (and also the [deployment link above](https://a9aeda25.next-on-pages-test-5h3.pages.dev/) seems to be working totally fine)

I did try to re-deploy the app multiple times (since in the past I saw this happen sporadically on deployments of the same app), none of the deployment presented the issue:
- https://2dddedbb.next-apps-for-testing.pages.dev/
- https://114d766a.next-apps-for-testing.pages.dev/
- https://be0239d9.next-apps-for-testing.pages.dev/
- https://8400719d.next-apps-for-testing.pages.dev/
- https://5a613c7d.next-apps-for-testing.pages.dev/
- https://8dcd64c0.next-apps-for-testing.pages.dev/
- https://e550956d.next-apps-for-testing.pages.dev/
- https://17cb6ed0.next-apps-for-testing.pages.dev/

So hopefully the problem should have completely gone away 🙂🤞

**Comment by tomaspietravallo** [comment]
Hey unfortunately I am facing a very similar (if not the same) issue. 

`npx next build` runs perfectly fine.. I suspect `@cloudflare/next-on-pages` is to blame..


```
▲  Traced Next.js server files in: 274.752ms
▲  Created all serverless functions in: 626.293ms
▲  Collected static files (public/, static/, .next/static): 4.955ms
▲  Build Completed in .vercel/output [12s]
⚡️ Completed `npx vercel build`.
⚡️ Unexpected error: ENOENT: no such file or directory, copyfile '/Users/tomaspietravallo/Documents/PROJECT/.vercel/output/functions/api/og.func/wasm/wasm_ef4866ecae192fd87727067cf2c0c0cf9fb8b020.wasm' -> '/Users/tomaspietravallo/Documents/PROJECT/.vercel/output/static/_worker.js/__next-on-pages-dist__/wasm/wasm_ef4866ecae192fd87727067cf2c0c0cf9fb8b020.wasm'
```

I am trying to use `@vercel/og` to generate images for og


```tsx
import React from "react";
import { ImageResponse } from "@cloudflare/pages-plugin-vercel-og/api";
import type { NextRequest } from "next/server";

export const runtime = 'edge';

export async function GET(request: NextRequest){
  return new ImageResponse(
    <Component/>,
    {
      width: 800,
      height: 400,
      status: 200,
      debug: false,
    }
  ) as Response;
};
```

**Comment by tomaspietravallo** [comment]
Ps the second link on the deployment you shared @dario-piotrowicz is not working "[Example of vercel/og usage (which relies on wasm)](https://a9aeda25.next-on-pages-test-5h3.pages.dev/api/og)" returns "internal server error for me"

**Comment by dario-piotrowicz** [comment]
> Ps the second link on the deployment you shared @dario-piotrowicz is not working "[Example of vercel/og usage (which relies on wasm)](https://a9aeda25.next-on-pages-test-5h3.pages.dev/api/og)" returns "internal server error" for me

@tomaspietravallo Thanks a bunch for letting me know 🙏 

This issue is quite bothersome as it has to be something runtime related... and it looks like the issue appears and disappears... making it very difficult to understand what's going wrong! 😓 
I'll have to try to see if anything can be done regarding it 😓

**Comment by dario-piotrowicz** [comment]
> Hey unfortunately I am facing a very similar (if not the same) issue.
> 
> `npx next build` runs perfectly fine.. I suspect `@cloudflare/next-on-pages` is to blame..

Yes I'd say that your issue is not the same... could you possibly create a [new issue](https://github.com/cloudflare/next-on-pages/issues/new?assignees=&labels=bug&projects=&template=bug-report.yml&title=%5B%F0%9F%90%9B+Bug%5D%3A+)? with also a minimal reproduction? 🙏

**Comment by tomaspietravallo** [comment]
I'm sure they did work. I hope you are able to find the cause, I'll be around to run any tests you might need help with. 

For whatever reason the custom font example on your site DOES work (im assuming old deployment/version) https://17cb6ed0.next-apps-for-testing.pages.dev/api/og-custom

I'll open a new issue, hopefully I'll get a minimal example ready by tomorrow

**Comment by Cahl-Dee** [comment]
Hey @dario-piotrowicz looks like the links discussed in this thread are working, but ours isn't 😓 

Seeing
```  "exceptions": [
    {
      "stack": "    at Object.ah (__next-on-pages-dist__/webpack/8ebb84e10b63f5ce7c8cbab62405cc23.js:56:31339)\n    at g (__next-on-pages-dist__/functions/src/middleware.func.js:1:2917)\n    at Module.ih (__next-on-pages-dist__/webpack/8ebb84e10b63f5ce7c8cbab62405cc23.js:2:141)\n    at Function.g (__next-on-pages-dist__/functions/src/middleware.func.js:1:2917)\n    at async Object.start (__next-on-pages-dist__/webpack/8ebb84e10b63f5ce7c8cbab62405cc23.js:56:33270)",
      "name": "ReferenceError",
      "message": "wasm_ef4866ecae192fd87727067cf2c0c0cf9fb8b020 is not defined",
      "timestamp": 1732297141542
    }
  ],```

When loading the OG Image. Curious if you had any suggestions on how to workaround the bug for now.

Thanks!

**Comment by dario-piotrowicz** [comment]
Hello 🙂 

Earlier this year we released our [Cloudflare adapter for OpenNext](https://opennext.js.org/cloudflare), and we have been continuously investing effort in its improvement since then. Existing deployments can continue to use the next-on-pages tooling, but the Cloudflare adapter for OpenNext is currently stable and recommended for deployment of Next.js applications to Cloudflare. 

As such, we will no longer be maintaining next-on-pages. We recently deprecated the package on npm and we plan to archive this repo on Monday Sep 29, 2025. In preparation for archival, we are closing all open issues and PRs. 

If you have concerns, please feel free to reply here before the repo is archived (after that, it will no longer be possible to comment) or to reach out to our team on [Discord](https://discord.com/invite/cloudflaredev).

Thanks so much for using and contributing to next-on-pages 🧡
